File: ios-app/BayitPlusApp/Models/WidgetModels.swift
import Foundation

// MARK: - Widget Enums

/// Type of widget (system-provided or user-created)
enum WidgetType: String, Decodable, Sendable {
    case system
    case personal
}

/// Content type a widget can display
enum WidgetContentType: String, Decodable, Sendable, CaseIterable {
    case liveChannel = "live_channel"
    case iframe
    case podcast
    case vod
    case radio
    case live
    case custom
    case audiobook

    /// SF Symbol icon name for this content type
    var iconName: String {
        switch self {
        case .liveChannel, .live: return "tv"
        case .podcast: return "mic"
        case .radio: return "radio"
        case .vod: return "film"
        case .audiobook: return "book"
        case .iframe: return "globe"
        case .custom: return "square.grid.2x2"
        }
    }

    /// Badge variant color for this content type
    var badgeVariant: String {
        switch self {
        case .liveChannel, .live: return "primary"
        case .podcast: return "success"
        case .radio: return "warning"
        case .vod, .audiobook: return "info"
        case .iframe: return "secondary"
        case .custom: return "primary"
        }
    }

    /// Display label for UI badges
    var displayLabel: String {
        switch self {
        case .liveChannel: return "Live Channel"
        case .iframe: return "Web"
        case .podcast: return "Podcast"
        case .vod: return "VOD"
        case .radio: return "Radio"
        case .live: return "Live"
        case .custom: return "Custom"
        case .audiobook: return "Audiobook"
        }
    }
}

// MARK: - Widget Position

/// Position and sizing of a widget on the dock grid
struct WidgetPosition: Decodable, Sendable {
    let x: Double
    let y: Double
    let width: Int
    let height: Int
    let zIndex: Int?
}

// MARK: - Widget Content

/// Content configuration within a widget
struct WidgetContent: Decodable, Sendable {
    let contentType: WidgetContentType?
    let liveChannelId: String?
    let podcastId: String?
    let contentId: String?
    let stationId: String?
    let audiobookId: String?
    let iframeUrl: String?
    let iframeTitle: String?
    let componentName: String?
}

// MARK: - Widget Item

/// A single widget belonging to the user
struct WidgetItem: Decodable, Sendable, Identifiable {
    let id: String
    let userId: String?
    let title: String
    let description: String?
    let type: WidgetType
    let systemWidgetId: String?
    let icon: String?
    let position: WidgetPosition?
    let content: WidgetContent?
    let isVisible: Bool?
    let isMinimized: Bool?
    let coverUrl: String?
    let createdAt: String?
    let updatedAt: String?
}

// MARK: - Available System Widget

/// A system widget that can be added, with its current subscription status
struct AvailableSystemWidget: Decodable, Sendable, Identifiable {
    let id: String
    let title: String
    let description: String?
    let icon: String?
    let coverUrl: String?
    let content: WidgetContent?
    let isAdded: Bool

    enum CodingKeys: String, CodingKey {
        case id, title, description, icon, content, isAdded
        case coverUrl = "cover_url"
    }
}

// MARK: - API Response Types

/// Response from GET /api/v1/widgets
struct WidgetsListResponse: Decodable, Sendable {
    let items: [WidgetItem]
    let total: Int
}

/// Response from GET /api/v1/widgets/system/available
struct AvailableSystemWidgetsResponse: Decodable, Sendable {
    let items: [AvailableSystemWidget]
    let total: Int
}

/// Response from POST /api/v1/widgets/system/{id}/add
struct WidgetActionResponse: Decodable, Sendable {
    let message: String
    let id: String?
    let widgetId: String?
}

/// Response from DELETE /api/v1/widgets/system/{id}/remove
struct WidgetDeleteResponse: Decodable, Sendable {
    let message: String
}

// MARK: - Widget Creation

/// Encodable content for creating a personal widget.
struct WidgetContentPayload: Encodable, Sendable {
    let contentType: String
    let liveChannelId: String?
    let podcastId: String?
    let contentId: String?
    let stationId: String?
    let iframeUrl: String?
    let iframeTitle: String?
}

/// Request body for POST /admin/widgets (personal widget creation).
struct CreateWidgetRequest: Encodable, Sendable {
    let title: String
    let description: String?
    let icon: String?
    let content: WidgetContentPayload
}

/// Response from POST /admin/widgets
struct CreateWidgetResponse: Decodable, Sendable {
    let id: String?
    let title: String?
    let message: String?
}

File: android-app/core/core-model/src/main/java/tv/bayit/plus/core/model/WidgetModels.kt
package tv.bayit.plus.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Type of widget (system-provided or user-created)
 */
@Serializable
enum class WidgetType {
    @SerialName("system")
    SYSTEM,

    @SerialName("personal")
    PERSONAL
}

/**
 * Content type a widget can display
 */
@Serializable
enum class WidgetContentType {
    @SerialName("live_channel")
    LIVE_CHANNEL,

    @SerialName("iframe")
    IFRAME,

    @SerialName("podcast")
    PODCAST,

    @SerialName("vod")
    VOD,

    @SerialName("radio")
    RADIO,

    @SerialName("live")
    LIVE,

    @SerialName("custom")
    CUSTOM,

    @SerialName("audiobook")
    AUDIOBOOK;

    val iconName: String
        get() = when (this) {
            LIVE_CHANNEL, LIVE -> "tv"
            PODCAST -> "mic"
            RADIO -> "radio"
            VOD -> "film"
            AUDIOBOOK -> "book"
            IFRAME -> "globe"
            CUSTOM -> "grid_view"
        }

    val displayLabel: String
        get() = when (this) {
            LIVE_CHANNEL -> "Live Channel"
            IFRAME -> "Web"
            PODCAST -> "Podcast"
            VOD -> "VOD"
            RADIO -> "Radio"
            LIVE -> "Live"
            CUSTOM -> "Custom"
            AUDIOBOOK -> "Audiobook"
        }
}

/**
 * Position and sizing of a widget on the dock grid
 */
@Serializable
data class WidgetPosition(
    val x: Double,
    val y: Double,
    val width: Int,
    val height: Int,
    @SerialName("z_index")
    val zIndex: Int? = null,
)

/**
 * Content configuration within a widget
 */
@Serializable
data class WidgetContent(
    @SerialName("content_type")
    val contentType: WidgetContentType? = null,
    @SerialName("live_channel_id")
    val liveChannelId: String? = null,
    @SerialName("podcast_id")
    val podcastId: String? = null,
    @SerialName("content_id")
    val contentId: String? = null,
    @SerialName("station_id")
    val stationId: String? = null,
    @SerialName("audiobook_id")
    val audiobookId: String? = null,
    @SerialName("iframe_url")
    val iframeUrl: String? = null,
    @SerialName("iframe_title")
    val iframeTitle: String? = null,
    @SerialName("component_name")
    val componentName: String? = null,
)

/**
 * A single widget belonging to the user
 */
@Serializable
data class WidgetItem(
    val id: String,
    @SerialName("user_id")
    val userId: String? = null,
    val title: String,
    val description: String? = null,
    val type: WidgetType,
    @SerialName("system_widget_id")
    val systemWidgetId: String? = null,
    val icon: String? = null,
    val position: WidgetPosition? = null,
    val content: WidgetContent? = null,
    @SerialName("is_visible")
    val isVisible: Boolean? = null,
    @SerialName("is_minimized")
    val isMinimized: Boolean? = null,
    @SerialName("cover_url")
    val coverUrl: String? = null,
    @SerialName("created_at")
    val createdAt: String? = null,
    @SerialName("updated_at")
    val updatedAt: String? = null,
)

/**
 * Response from GET /api/v1/widgets
 */
@Serializable
data class WidgetsListResponse(
    val items: List<WidgetItem>,
    val total: Int,
)
